src/controllers/hid/hidiooutputreport.cpp

```diff
@@ -0,0 +1,155 @@
+    m_cachedData.append(reportId);
+    m_lastSentData.reserve(kReportIdSize + reportDataSize);
+    m_lastSentData.append(reportId);
+}
+
+void HidIoOutputReport::updateCachedData(const QByteArray& data,
+        const mixxx::hid::DeviceInfo& deviceInfo,
+        const RuntimeLoggingCategory& logOutput) {
+    auto cacheLock = lockMutex(&m_cachedDataMutex);
+
+    if (!m_lastCachedDataSize) {
+        // First call updateCachedData for this report
+        m_lastCachedDataSize = data.size();
+
+    } else {
+        if (m_possiblyUnsentDataCached) {
+            qCDebug(logOutput) << "t:" << mixxx::Time::elapsed().formatMillisWithUnit()
+                               << "Skipped superseded OutputReport"
+                               << deviceInfo.formatName() << "serial #"
+                               << deviceInfo.serialNumber() << "(Report ID"
+                               << m_reportId << ")";
+        }
+
+        // The size of an HID report is defined in a HID device and can't vary at runtime
+        if (m_lastCachedDataSize != data.size()) {
+            qCWarning(logOutput) << "Size of report (with Report ID"
+                                 << m_reportId << ") changed from"
+                                 << m_lastCachedDataSize
+                                 << "to" << data.size() << "for"
+                                 << deviceInfo.formatName() << "serial #"
+                                 << deviceInfo.serialNumber()
+                                 << "- This indicates a bug in the mapping code!";
+            m_lastCachedDataSize = data.size();
+        }
+    }
+
+    // Deep copy with reusing the already allocated heap memory
+    // The first byte with the ReportID is not overwritten
+    qByteArrayReplaceWithPositionAndSize(&m_cachedData,
+            kReportIdSize,
+            m_cachedData.size(),
+            data.constData(),
+            data.size());
+    m_possiblyUnsentDataCached = true;
+}
+
+bool HidIoOutputReport::sendCachedData(QMutex* pHidDeviceAndPollMutex,
+        hid_device* pHidDevice,
+        const mixxx::hid::DeviceInfo& deviceInfo,
+        const RuntimeLoggingCategory& logOutput) {
+    auto startOfHidWrite = mixxx::Time::elapsed();
+
+    auto cacheLock = lockMutex(&m_cachedDataMutex);
+
+    if (!m_possiblyUnsentDataCached) {
+        // Return with false, to signal the caller, that no time consuming IO operation was necessary
+        return false;
+    }
+
+    if (!m_lastSentData.compare(m_cachedData)) {
+        // An HID OutputReport can contain only HID OutputItems.
+        // HID OutputItems are defined to represent the state of one or more similar controls or LEDs.
+        // Only HID Feature items may be attributes of other items.
+        // This means there is always a one to one relationship to the state of control(s)/LED(s),
+        // and if the state is not changed, there's no need to execute the time consuming hid_write again.
```

Comment:
This is incorrect and has broken the output messages driving the motor in my Kontrol S4 Mk3 mapping. The motor stops turning if no output packets are sent. To keep a constant velocity, identical output packets must be sent repeatedly.

Reply:
I can think of three ways forward:
1. Remove this optimization
2. Add an optional parameter to the `send` function to opt in to allowing duplicate packets.
3. Add a separate function like `sendAllowDuplicates` for this

Reply:
I will look how to solve this!

Reply:
The default should behavior should ignore duplicate packets. The timing and jitter of the packets is unpredictable and usually not suitable for implementing a real-time control cycle that requires exact cycle times.

I vote for adding an extra parameter to override the default behavior if needed.

Reply:
I would be fine with solution 2. too!

But I could also imagine fourth solution:
4. Add a function configureOutputReport(int ReportID, bool sendDuplicates = false) which needs to be called only once in the init section of the JS mapping code.

Reply:
I would prefer option 2 to add an optional parameter to the existing `send` function rather than add more functions for this edge case.

Reply:
I recommend to avoid stateful code if possible. Let the script pass an additional parameter on each invocation is less error prone and provides more flexibility.

Reply:
Ok, than I will provide a PR for 2.